File: src/prod/src/ServiceModel/remotereplicatoracknowledgementstatus.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace std;
using namespace Common;
using namespace ServiceModel;

RemoteReplicatorAcknowledgementStatus::RemoteReplicatorAcknowledgementStatus()
    : replicationStreamAcknowledgementDetail_()
    , copyStreamAcknowledgementDetail_()
{
}

RemoteReplicatorAcknowledgementStatus::RemoteReplicatorAcknowledgementStatus(
    RemoteReplicatorAcknowledgementDetail && replicationStreamAcknowledgementDetail,
    RemoteReplicatorAcknowledgementDetail && copyStreamAcknowledgementDetail)
    : replicationStreamAcknowledgementDetail_(move(replicationStreamAcknowledgementDetail))
    , copyStreamAcknowledgementDetail_(move(copyStreamAcknowledgementDetail))
{
}

RemoteReplicatorAcknowledgementStatus::RemoteReplicatorAcknowledgementStatus(RemoteReplicatorAcknowledgementStatus && other)
    : replicationStreamAcknowledgementDetail_(move(other.replicationStreamAcknowledgementDetail_))
    , copyStreamAcknowledgementDetail_(move(other.copyStreamAcknowledgementDetail_))
{
}

RemoteReplicatorAcknowledgementStatus & RemoteReplicatorAcknowledgementStatus::operator=(RemoteReplicatorAcknowledgementStatus && other)
{
    if (this != &other)
    {
        this->replicationStreamAcknowledgementDetail_ = move(other.replicationStreamAcknowledgementDetail_);
        this->copyStreamAcknowledgementDetail_ = move(other.copyStreamAcknowledgementDetail_);
    }

    return *this;
}

ErrorCode RemoteReplicatorAcknowledgementStatus::ToPublicApi(
    __in Common::ScopedHeap & heap, 
    __out FABRIC_REMOTE_REPLICATOR_ACKNOWLEDGEMENT_STATUS & publicResult) const 
{
    auto replicationAckDetails = heap.AddItem<FABRIC_REMOTE_REPLICATOR_ACKNOWLEDGEMENT_DETAIL>();
    auto copyAckDetails = heap.AddItem<FABRIC_REMOTE_REPLICATOR_ACKNOWLEDGEMENT_DETAIL>();

    publicResult.ReplicationStreamAcknowledgementDetails = replicationAckDetails.GetRawPointer();
    publicResult.CopyStreamAcknowledgementDetails = copyAckDetails.GetRawPointer();

    this->replicationStreamAcknowledgementDetail_.ToPublicApi(heap, *publicResult.ReplicationStreamAcknowledgementDetails);
    this->copyStreamAcknowledgementDetail_.ToPublicApi(heap, *publicResult.CopyStreamAcknowledgementDetails);

    return ErrorCode::Success();
}

ErrorCode RemoteReplicatorAcknowledgementStatus::FromPublicApi(
    __in FABRIC_REMOTE_REPLICATOR_ACKNOWLEDGEMENT_STATUS & publicResult)
{
    this->replicationStreamAcknowledgementDetail_.FromPublicApi(*publicResult.ReplicationStreamAcknowledgementDetails);
    this->copyStreamAcknowledgementDetail_.FromPublicApi(*publicResult.CopyStreamAcknowledgementDetails);

    return ErrorCode();
}

void RemoteReplicatorAcknowledgementStatus::WriteTo(Common::TextWriter& w, Common::FormatOptions const &) const
{
    w.Write(ToString());
}

std::wstring RemoteReplicatorAcknowledgementStatus::ToString() const
{
    return wformatString("\nReplicationStreamAcknowledgementDetails = {0}\nCopyStreamAcknowledgementDetails = {1}\n",
        this->replicationStreamAcknowledgementDetail_,
        this->copyStreamAcknowledgementDetail_);
}

File: src/prod/src/managed/Api/src/System/Fabric/query/RemoteReplicatorAcknowledgementStatus.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric.Query
{
    using System.Fabric.Interop;

    /// <summary>
    /// Provides copy and replication stream acknowledgement details. Member of <see cref="System.Fabric.Query.RemoteReplicatorStatus" />
    /// </summary>
    public sealed class RemoteReplicatorAcknowledgementStatus
    {
        /// <summary>
        /// <para>Initializes a new instance of the <see cref="System.Fabric.Query.RemoteReplicatorAcknowledgementStatus" /> class.</para>
        /// </summary>
        public RemoteReplicatorAcknowledgementStatus()
        {
        }

        /// <summary>
        /// Contains details regarding acknowledgement for the replication operation stream
        /// </summary>
        /// <value>
        /// <para>Returns RemoteReplicatorAcknowledgementDetail object for replication operation stream. See <see cref="System.Fabric.Query.RemoteReplicatorAcknowledgementDetail" /> for more information. </para>
        /// </value>
        public RemoteReplicatorAcknowledgementDetail ReplicationStreamAcknowledgementDetail { get; private set; }

        /// <summary>
        /// Contains details regarding acknowledgement for the copy operation stream
        /// </summary>
        /// <value>
        /// <para>Returns RemoteReplicatorAcknowledgementDetail object for copy operation stream. See <see cref="System.Fabric.Query.RemoteReplicatorAcknowledgementDetail" /> for more information.</para>
        /// </value>
        public RemoteReplicatorAcknowledgementDetail CopyStreamAcknowledgementDetail { get; private set; }

        internal static unsafe RemoteReplicatorAcknowledgementStatus CreateFromNative(NativeTypes.FABRIC_REMOTE_REPLICATOR_ACKNOWLEDGEMENT_STATUS nativeResultItem)
        {
            return new RemoteReplicatorAcknowledgementStatus
            {
                ReplicationStreamAcknowledgementDetail =
                    RemoteReplicatorAcknowledgementDetail.CreateFromNative(*(NativeTypes.FABRIC_REMOTE_REPLICATOR_ACKNOWLEDGEMENT_DETAIL*)nativeResultItem.ReplicationStreamAcknowledgementDetails),
                CopyStreamAcknowledgementDetail = 
                    RemoteReplicatorAcknowledgementDetail.CreateFromNative(*(NativeTypes.FABRIC_REMOTE_REPLICATOR_ACKNOWLEDGEMENT_DETAIL*)
                    nativeResultItem.CopyStreamAcknowledgementDetails)
            };
        }
    }
}